### PR TITLE
RFC 0006-hardening-against-toctou

### PR DESCRIPTION
**Pull request by bsimondi**

### CODE REVIEW

File: docs/rfcs/accepted/0000-hardening-against-toctou/README.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+## Motivation / use-cases
+
+TOCTOU is a class of security vulnerabilities that is hard to avoid in complex code bases, and can 
+be especially challenging when exchanging data between different execution environment with shared 
+memories.
+All parameters and fields coming from the outside world (anything calling inside of the code) 
+should be sanitized, but special concern is to make sure there is no possibilities for the value 
+to change between the sanitization and the use of the value. Even a very short window can present 
+opportunities and be abused. I general is it preferable to cache (create a copy of) all structures
```
Comment (ericeberry) [inline docs/rfcs/accepted/0000-hardening-against-toctou/README.md:19]:
```suggestion
opportunities and be abused. In general is it preferable to cache (create a copy of) all structures 
```

**Review by ericeberry — CHANGES_REQUESTED** [review]
Please change directory to 0006-hardening-against-toctou

**Review by kihun-amlogic — APPROVED** [review]

**Review by LisShen — APPROVED** [review]

**Review by hchoi-brcm — DISMISSED** [review]

**Review by ericeberry — APPROVED** [review]

File: docs/rfcs/accepted/0006-hardening-against-toctou/README.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+- Leader(s): Eric Berry
+
+## Introduction
+
+Add mechanics to harden the code against Time-of-Check / Time-of-Use (TOCTOU) attacks.
+
+## Motivation / use-cases
+
+TOCTOU is a class of security vulnerabilities that is hard to avoid in complex code bases, and can 
+be especially challenging when exchanging data between different execution environment with shared
```
Comment (mhabrat) [inline docs/rfcs/accepted/0006-hardening-against-toctou/README.md:14]:
```suggestion
be especially challenging when exchanging data between different execution environments with shared 
```

File: docs/rfcs/accepted/0006-hardening-against-toctou/README.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+## Introduction
+
+Add mechanics to harden the code against Time-of-Check / Time-of-Use (TOCTOU) attacks.
+
+## Motivation / use-cases
+
+TOCTOU is a class of security vulnerabilities that is hard to avoid in complex code bases, and can 
+be especially challenging when exchanging data between different execution environment with shared 
+memories.
+All parameters and fields coming from the outside world (anything calling inside of the code)
```
Comment (mhabrat) [inline docs/rfcs/accepted/0006-hardening-against-toctou/README.md:16]:
```suggestion
All parameters and fields coming from the outside world (anything calling inside of the `tasecureapi` code) 
```

File: docs/rfcs/accepted/0006-hardening-against-toctou/README.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+## Motivation / use-cases
+
+TOCTOU is a class of security vulnerabilities that is hard to avoid in complex code bases, and can 
+be especially challenging when exchanging data between different execution environment with shared 
+memories.
+All parameters and fields coming from the outside world (anything calling inside of the code) 
+should be sanitized, but special concern is to make sure there is no possibilities for the value 
+to change between the sanitization and the use of the value. Even a very short window can present 
+opportunities and be abused. In general is it preferable to cache (create a copy of) all structures
```
Comment (mhabrat) [inline docs/rfcs/accepted/0006-hardening-against-toctou/README.md:19]:
```suggestion
opportunities and be abused. In general it is preferable to cache (create a copy of) all structures 
```

File: docs/rfcs/accepted/0006-hardening-against-toctou/README.md [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+## Motivation / use-cases
+
+TOCTOU is a class of security vulnerabilities that is hard to avoid in complex code bases, and can 
+be especially challenging when exchanging data between different execution environment with shared 
+memories.
+All parameters and fields coming from the outside world (anything calling inside of the code) 
+should be sanitized, but special concern is to make sure there is no possibilities for the value 
+to change between the sanitization and the use of the value. Even a very short window can present 
+opportunities and be abused. In general is it preferable to cache (create a copy of) all structures 
+and parameters coming from outside of the code.
```
Comment (mhabrat) [inline docs/rfcs/accepted/0006-hardening-against-toctou/README.md:20]:
```suggestion
and parameters coming from outside of the `tasecureapi` code.
```

File: docs/rfcs/accepted/0006-hardening-against-toctou/README.md [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+Abuse of TOCTOU attacks can lead to serious security vulnerabilities directly or indirectly 
+leading to data compromise or arbitrary data execution in the worse cases.
```
Comment (mhabrat) [inline docs/rfcs/accepted/0006-hardening-against-toctou/README.md:22]:
```suggestion
leading to data compromise or arbitrary data execution in the worst cases.
```

File: docs/rfcs/accepted/0006-hardening-against-toctou/README.md [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+## Updates/Obsoletes
+
+-
```
Comment (mhabrat) [inline docs/rfcs/accepted/0006-hardening-against-toctou/README.md:26]: Recommend using "None" in all blank sections.

File: docs/rfcs/accepted/0006-hardening-against-toctou/README.md [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+## Affected platforms
+
+-
```
Comment (mhabrat) [inline docs/rfcs/accepted/0006-hardening-against-toctou/README.md:30]: All?

File: docs/rfcs/accepted/0006-hardening-against-toctou/README.md [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+## Open Source Dependencies
+
+-
+
+## Detailed design
+
+Code can rapidly get complex and even with the best intent and expertize will eventually introduce 
+mistakes leading to TOCTOU opportunities.
+
+The recommendation is to systematically copy all data coming from the outside world at the
```
Comment (mhabrat) [inline docs/rfcs/accepted/0006-hardening-against-toctou/README.md:41]:
```suggestion
The recommendation is to systematically copy all data coming from the outside world with the 
```

File: docs/rfcs/accepted/0006-hardening-against-toctou/README.md [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+there are such practices, then the nested pointers should obey the same rules and the pointed 
+structure shall be cached as well.
+
+## Drawbacks
+
+-
+
+## Alternatives considered
+
+-
```
Comment (mhabrat) [inline docs/rfcs/accepted/0006-hardening-against-toctou/README.md:59]:
```suggestion
- None.

```

File: docs/rfcs/accepted/0006-hardening-against-toctou/README.md [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
+## Future possibilities
+
+-
+
+## References
+
+-
```
Comment (mhabrat) [inline docs/rfcs/accepted/0006-hardening-against-toctou/README.md:70]:
```suggestion
- [TOCTOU Issue](https://cwe.mitre.org/data/definitions/367.html).
```

File: docs/rfcs/accepted/0006-hardening-against-toctou/README.md [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+
+The recommendation is to systematically copy all data coming from the outside world at the 
+exception of video or audio payloads. 
+Rules could be created to reserve some of the RPC Global Platform (GP) parameters for 
+parameters/values, and some for payloads, so that some buffers are always copied/cache internally 
+(e.g. params[] 0 and 1), and some could be used directly even if shared but will never be involved 
+in code logic decisions (e.g. params[] 2 and 3). Apply this consistent protocol for all of the 
+commands that can be dispatched, at the dispatcher level.
+
+Note that, particular attention should be made to eventually nested pointers (discouraged), if
```
Comment (mhabrat) [inline docs/rfcs/accepted/0006-hardening-against-toctou/README.md:49]:
```suggestion
Note that, particular attention should be made to protect nested pointers. While the use of nested pointers is discouraged, if 
```

**Review by mhabrat — CHANGES_REQUESTED** [review]

File: docs/rfcs/accepted/0006-hardening-against-toctou/README.md (hunk 8 above)
Comment (bsimondi) [reply docs/rfcs/accepted/0006-hardening-against-toctou/README.md:30]: Yes, I guess. I am open to debate, which platforms would not be affected?

**Review by ericeberry — DISMISSED** [review]

File: docs/rfcs/accepted/0006-hardening-against-toctou/README.md (hunk 10 above)
Comment (ericeberry) [reply docs/rfcs/accepted/0006-hardening-against-toctou/README.md]: Done

File: docs/rfcs/accepted/0006-hardening-against-toctou/README.md (hunk 8 above)
Comment (ericeberry) [reply docs/rfcs/accepted/0006-hardening-against-toctou/README.md]: Done

File: docs/rfcs/accepted/0006-hardening-against-toctou/README.md (hunk 7 above)
Comment (ericeberry) [reply docs/rfcs/accepted/0006-hardening-against-toctou/README.md]: Done

File: docs/rfcs/accepted/0006-hardening-against-toctou/README.md [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+
+The recommendation is to systematically copy all data coming from the outside world with the 
+exception of video or audio payloads. 
+Rules could be created to reserve some of the RPC Global Platform (GP) parameters for 
+parameters/values, and some for payloads, so that some buffers are always copied/cache internally 
+(e.g. params[] 0 and 1), and some could be used directly even if shared but will never be involved 
+in code logic decisions (e.g. params[] 2 and 3). Apply this consistent protocol for all of the 
+commands that can be dispatched, at the dispatcher level.
+
+Note that, particular attention should be made to protect nested pointers. While the use of nested pointers is discouraged, if
```
Comment (mhabrat) [inline docs/rfcs/accepted/0006-hardening-against-toctou/README.md:49]: Looks like line length is too long here.

File: docs/rfcs/accepted/0006-hardening-against-toctou/README.md [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+in code logic decisions (e.g. params[] 2 and 3). Apply this consistent protocol for all of the 
+commands that can be dispatched, at the dispatcher level.
+
+Note that, particular attention should be made to protect nested pointers. While the use of nested pointers is discouraged, if 
+there are such practices, then the nested pointers should obey the same rules and the pointed 
+structure shall be cached as well.
+
+## Drawbacks
+
+None
```
Comment (mhabrat) [inline docs/rfcs/accepted/0006-hardening-against-toctou/README.md:55]: After further consideration, I think we have the potential for 2 drawbacks:
1. Performance based on the number of copies and amount of data
2. TEE memory resource limitations based on the amount and quantity of copies being maintained in the TEE.
Any idea what our worst case scenario is in the current implementation and the corresponding impact with respect to these drawbacks?

**Review by mhabrat — CHANGES_REQUESTED** [review]

**Review by mhabrat — APPROVED** [review]

File: docs/rfcs/accepted/0006-hardening-against-toctou/README.md (hunk 14 above)
Comment (ericeberry) [reply docs/rfcs/accepted/0006-hardening-against-toctou/README.md]: No we don't know. The observed impact was due to memory checks not data copying. Most copied data due to this RFC is < 100 bytes. Anything more is passed through references.

File: docs/rfcs/accepted/0006-hardening-against-toctou/README.md (hunk 13 above)
Comment (ericeberry) [reply docs/rfcs/accepted/0006-hardening-against-toctou/README.md]: Done

**Pull request merged**